### PR TITLE
feat(runtime): Implement `console.Console`

### LINKED ISSUE #3625: Implement `console.Console` constructor.
### What version of Bun is running?

0.6.13

### What platform is your computer?

Linux 5.15.49-linuxkit-pr aarch64 aarch64

### What steps can reproduce the bug?

Execute this code in bun
```js
const log = new console.Console();
```
 

### What is the expected behavior?

Should not fail to call `new console.Console()` and must create a new Console instances.

### What do you see instead?

```
const log = new console.Console();
                ^
TypeError: undefined is not a constructor (evaluating 'new console.Console')
```

### Additional information

This method is implemented in nodejs and some npm packages use it which causes it to fail when using it with bun.

Method documentation for nodejs https://nodejs.org/api/console.html#new-consoleoptions

### LINKED ISSUE #4610: cucumber fails to run
### What version of Bun is running?

1.0.0+822a00c4d508b54f650933a73ca5f4a3af9a7983

### What platform is your computer?

Linux 5.15.90 x86_64 unknown

### What steps can reproduce the bug?

```sh
$ bun add @cucumber/cucumber
# ...
$ bun --bun run cucumber-js
TypeError: undefined is not a constructor (evaluating 'new console_1.Console(this.stream)')
    at new ConsoleLogger (/app/upm/tests/node_modules/@cucumber/cucumber/lib/api/console_logger.js:8:56)
    at <anonymous> (/app/upm/tests/node_modules/@cucumber/cucumber/lib/api/load_configuration.js:21:8)
    at loadConfiguration (/app/upm/tests/node_modules/@cucumber/cucumber/lib/api/load_configuration.js:17:8)
    at <anonymous> (/app/upm/tests/node_modules/@cucumber/cucumber/lib/cli/index.js:46:26)
    at run (/app/upm/tests/node_modules/@cucumber/cucumber/lib/cli/index.js:19:14)
    at <anonymous> (/app/upm/tests/node_modules/@cucumber/cucumber/lib/cli/run.js:31:4)
    at run (/app/upm/tests/node_modules/@cucumber/cucumber/lib/cli/run.js:16:8)
    at <anonymous> (/app/upm/tests/node_modules/@cucumber/cucumber/bin/cucumber.js:2:37)
    at global code (/app/upm/tests/node_modules/@cucumber/cucumber/bin/cucumber.js:3:8)
error: "cucumber-js" exited with code 1 (SIGHUP)
```

### What is the expected behavior?

i expect the behavior with `--bun` to match this behavior, which is actually using node:

```sh
$ bun run cucumber-js      


0 scenarios
0 steps
0m00.000s (executing steps: 0m00.000s)
```

### LINKED ISSUE #4737: gulp not running due to node:Console
### What version of Bun is running?

1.0.0+822a00c4d508b54f650933a73ca5f4a3af9a7983

### What platform is your computer?

Linux 6.4.12-arch1-1 x86_64 unknown

### What steps can reproduce the bug?

- bun add gulp
- bun install
- bun run gulp

### What is the expected behavior?

gulp running like it would on nodejs

### What do you see instead?

```
➜ bun run gulp
15 | 
16 | // Needed to add this because node 10 decided to start coloring log output randomly
17 | var console;
18 | if (nodeVersion.major >= 10) {
19 |   // Node 10 also changed the way this is constructed
20 |   console = new Console({
                ^
TypeError: undefined is not a constructor (evaluating 'new Console({
      stdout: process.stdout,
      stderr: process.stderr,
      colorMode: !1
    })')
      at /home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/fancy-log/index.js:20:12
      at globalThis (/home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/fancy-log/index.js:92:23)
      at require (:1:20)
      at /home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/gulp-cli/lib/shared/log/to-console.js:3:4
      at globalThis (/home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/gulp-cli/lib/shared/log/to-console.js:58:17)
      at require (:1:20)
      at /home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/gulp-cli/index.js:20:4
      at globalThis (/home/noraj/Dev/rawsec-cybersecurity-inventory/node_modules/gulp-cli/index.js:111:17)
```

### Additional information

Ref. https://bun.sh/docs/runtime/nodejs-apis

> If you run into any bugs with a particular package, please [open an issue](https://bun.sh/issues). Opening issues for compatibility bugs helps us prioritize what to work on next.

Seems due at least to [node:console](https://nodejs.org/api/console.html) incomplete support.

Asking for bun support on upstream https://github.com/gulpjs/gulp/issues/2737

### LINKED ISSUE #4898: Error  undefined is not a constructor (evaluating 'new console_1.Console(process.stderr)')
### What version of Bun is running?

1.0

### What platform is your computer?

Darwin 22.5.0 arm64 arm

### What steps can reproduce the bug?

- using docker images oven/bun with nestjs when bun run build show "Error  undefined is not a constructor (evaluating 'new console_1.Console(process.stderr)')"

### What is the expected behavior?

_No response_

### What do you see instead?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by paperclover**
### What does this PR do?

Copies node's implementation for `console.Console`, which gives the console interface but piped to any node stream. This enables:

- Closes #3625
- MightFix #4145
- Closes #4610
- Closes #4898
- Closes #4737
- Helps #2034

```ts
import { Console } from "console";
import { createWriteStream } from "fs";

const writer = new Console({ stdout: createWriteStream("log.txt") });

writer.log("hello");
writer.log("world", { x: 2 });
```

and this will write to the file

```
hello
world { x: 2 }
```

TODO:
- [ ] Tests.
- [ ] Bug with printing will reveal the options object somehow.
- [ ] Self-Review.

### CODE REVIEW

File: src/js/node/util.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ var formatRegExp = /%[sdjfoc%]/g;
 function formatWithOptions(inspectOptions, f) {
   if (!isString(f)) {
     var objects = [];
-    for (var i = 0; i < arguments.length; i++) {
+    for (var i = 1; i < arguments.length; i++) {
```
Comment (paperclover) [inline src/js/node/util.js:30]: (fixes `formatWithOptions({colors:true}, {x:2)`)

File: src/js/builtins/ConsoleObject.ts [hunk 2]
```diff
@@ -122,3 +122,763 @@ export function write(this: Console, input) {
...
+        (code >= 0x1f200 && code <= 0x1f251) ||
+        // Miscellaneous Symbols and Pictographs 0x1f300 - 0x1f5ff
+        // Emoticons 0x1f600 - 0x1f64f
+        (code >= 0x1f300 && code <= 0x1f64f) ||
+        // CJK Unified Ideographs Extension B .. Tertiary Ideographic Plane
+        (code >= 0x20000 && code <= 0x3fffd))
+    );
+  };
+
+  var isZeroWidthCodePoint = code => {
```
Comment (Jarred-Sumner) [inline src/js/builtins/ConsoleObject.ts:210]: really wish this was like 95% less javascript

File: src/js/builtins/ConsoleObject.ts [hunk 3]
```diff
@@ -122,3 +122,763 @@ export function write(this: Console, input) {
...
+
+  const kSecond = 1000;
+  const kMinute = 60 * kSecond;
+  const kHour = 60 * kMinute;
+
+  // Regex used for ansi escape code splitting
+  // Adopted from https://github.com/chalk/ansi-regex/blob/HEAD/index.js
+  // License: MIT, authors: @sindresorhus, Qix-, arjunmehta and LitoMore
+  // Matches all ansi escape code sequences in a string
+  var ansiPattern =
```
Comment (Jarred-Sumner) [inline src/js/builtins/ConsoleObject.ts:160]: this is going to be extremely slow due to interpretor overhead

File: src/bun.js/bindings/ZigGlobalObject.cpp [hunk 4]
```diff
@@ -3317,7 +3317,45 @@ JSValue getEventSourceConstructor(VM& vm, JSObject* thisObject)
         throwException(globalObject, scope, returnedException.get());
     }
 
-    RELEASE_AND_RETURN(scope, result);
+    RETURN_IF_EXCEPTION(scope, {});
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ZigGlobalObject.cpp]: i think i messed the merge up

File: src/js/builtins/ConsoleObject.ts (hunk 3 above)
Comment (melroy89) [reply src/js/builtins/ConsoleObject.ts:160]: but still.. it's merged?

File: src/js/builtins/ConsoleObject.ts (hunk 2 above)
Comment (melroy89) [reply src/js/builtins/ConsoleObject.ts:210]: I hear you.

**Pull request merged**